From 5815ce59747fac4fa5f87ffcd0c4f5b8311221de Mon Sep 17 00:00:00 2001
From: Daniel Chavolla <daniel.chavolla@windriver.com>
Date: Sat, 22 Oct 2016 20:22:29 -0400
Subject: [PATCH] pass down image_ref to compute's rebuild instance

The numa topology info set into the image metadata, was being lost after
an evacuation.
The reason for this issue is that the original image info was not been passed
to compute.manager.ComputeManager.rebuild_instance by the evacuation code in
compute.api.API.evacuate. This caused the numa topology info to be ignored when
the rebuild_claim object was made.

(cherry picked from commit e26be3a)

addendum to commit e26be3a39e2 to fix rebuild to a new image case

compute.manager.ComputeManager.rebuild_instance to take new image_ref
when passed down by a rebuild.

(cherry picked from commit a47b686)
---
 nova/compute/api.py                     | 2 +-
 nova/compute/manager.py                 | 7 +++++--
 nova/tests/unit/compute/test_compute.py | 2 +-
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/nova/compute/api.py b/nova/compute/api.py
index 19bb91f..592b660 100644
--- a/nova/compute/api.py
+++ b/nova/compute/api.py
@@ -3952,7 +3952,7 @@ class API(base.Base):
                        new_pass=admin_password,
                        injected_files=None,
                        image_ref=None,
-                       orig_image_ref=None,
+                       orig_image_ref=instance.image_ref,
                        orig_sys_metadata=None,
                        bdms=None,
                        recreate=True,
diff --git a/nova/compute/manager.py b/nova/compute/manager.py
index cd538c5..9a7e53e 100644
--- a/nova/compute/manager.py
+++ b/nova/compute/manager.py
@@ -2805,8 +2805,11 @@ class ComputeManager(manager.Manager):
             rebuild_claim = claims.NopClaim
 
         image_meta = {}
-        if image_ref:
-            image_meta = self.image_api.get(context, image_ref)
+        if image_ref or orig_image_ref:
+            if image_ref:
+                image_meta = self.image_api.get(context, image_ref)
+            else:
+                image_meta = self.image_api.get(context, orig_image_ref)
 
         # NOTE(mriedem): On a recreate (evacuate), we need to update
         # the instance's host and node properties to reflect it's
diff --git a/nova/tests/unit/compute/test_compute.py b/nova/tests/unit/compute/test_compute.py
index d9a5e5c..298dc12 100644
--- a/nova/tests/unit/compute/test_compute.py
+++ b/nova/tests/unit/compute/test_compute.py
@@ -10202,7 +10202,7 @@ class ComputeAPITestCase(BaseTestCase):
                 new_pass=None,
                 injected_files=None,
                 image_ref=None,
-                orig_image_ref=None,
+                orig_image_ref=instance.image_ref,
                 orig_sys_metadata=None,
                 bdms=None,
                 recreate=True,
-- 
2.7.4

